File: spring-web-java/src/main/java/spring/web/java/interfaces/controller/ItemController.java
package spring.web.java.interfaces.controller;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import spring.web.java.application.service.ItemService;
import spring.web.java.interfaces.dto.request.ItemSaveRequest;
import spring.web.java.interfaces.dto.response.ItemResponse;

@RestController
@RequestMapping("/items")
@RequiredArgsConstructor
public class ItemController {

	private final ItemService itemService;

	@PostMapping
	@ResponseStatus(HttpStatus.CREATED)
	public ItemResponse saveItem(@RequestBody @Valid ItemSaveRequest itemSaveRequest) {
		return itemService.saveItem(itemSaveRequest);
	}

	@GetMapping
	public Page<ItemResponse> findItems(Pageable pageable, @RequestParam(required = false) String name) {
		return itemService.findItems(pageable, name);
	}

	@GetMapping("/{id}")
	public ItemResponse findItem(@PathVariable Long id) {
		return itemService.findItem(id);
	}

	@PutMapping("/{id}")
	public ItemResponse putItem(@PathVariable Long id, @RequestBody @Valid ItemSaveRequest itemSaveRequest) {
		return itemService.putItem(id, itemSaveRequest);
	}

	@DeleteMapping("/{id}")
	@ResponseStatus(HttpStatus.NO_CONTENT)
	public void deleteItem(@PathVariable Long id) {
		itemService.deleteItem(id);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/presentation/controller/ItemController.kt
package spring.web.kotlin.presentation.controller

import org.springframework.data.domain.Pageable
import org.springframework.data.web.PageableDefault
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.security.access.prepost.PreAuthorize
import org.springframework.validation.annotation.Validated
import org.springframework.web.bind.annotation.*
import spring.web.kotlin.application.service.ItemService
import spring.web.kotlin.presentation.dto.request.ItemSaveRequest
import spring.web.kotlin.presentation.dto.response.ItemResponse

@RestController
@RequestMapping("/items")
class ItemController(
    private val itemService: ItemService
) {
    @PostMapping
    @PreAuthorize("hasAuthority('ITEM_WRITER')")
    @ResponseStatus(HttpStatus.CREATED)
    fun saveItem(@RequestBody @Validated itemSaveRequest: ItemSaveRequest) = itemService.saveItem(itemSaveRequest)

    @GetMapping
    @PreAuthorize("hasAuthority('ITEM_READER')")
    fun findItems(@PageableDefault pageable: Pageable, @RequestParam(required = false) name: String?) =
        itemService.findItems(pageable, name)

    @GetMapping("/{id}")
    @PreAuthorize("hasAuthority('ITEM_READER')")
    fun findItem(@PathVariable id: Long) = itemService.findItem(id)

    @PutMapping("/{id}")
    @PreAuthorize("hasAuthority('ITEM_WRITER')")
    fun putItem(
        @PathVariable id: Long,
        @RequestBody @Validated itemSaveRequest: ItemSaveRequest
    ): ResponseEntity<ItemResponse> {
        val (isNew, itemResponse) = itemService.putItem(id, itemSaveRequest)
        val status = if (isNew) HttpStatus.CREATED else HttpStatus.OK

        return ResponseEntity.status(status).body(itemResponse)
    }

    @DeleteMapping("/{id}")
    @PreAuthorize("hasAuthority('ITEM_WRITER')")
    @ResponseStatus(HttpStatus.NO_CONTENT)
    fun deleteItem(@PathVariable id: Long) {
        itemService.deleteItem(id)
    }
}